pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/periodictask/ControllerPeriodicTask.java

```diff
@@ -31,152 +31,90 @@
-    process(tableNamesWithType);
-    LOGGER.info("Finish processing {} tables in periodic task: {} in {}ms", numTables, getTaskName(),
-        (System.currentTimeMillis() - startTime));
-
-    _periodicTaskInProgress = false;
-  }
-
-  /**
-   * Stops the ControllerPeriodicTask by enabling the _stopPeriodicTask flag. The flag ensures that processing of no new table begins.
-   * This method waits for the in progress ControllerPeriodicTask to finish the table being processed, until MAX_CONTROLLER_PERIODIC_TASK_STOP_TIME_MILLIS
-   * Finally, it invokes the stopTask for any specific cleanup at the individual task level
-   */
-  @Override
-  public final void stop() {
-    _stopPeriodicTask = true;
-
-    LOGGER.info("Waiting for periodic task {} to finish, maxWaitTimeMillis = {}", getTaskName(),
-        MAX_CONTROLLER_PERIODIC_TASK_STOP_TIME_MILLIS);
-    long millisToWait = MAX_CONTROLLER_PERIODIC_TASK_STOP_TIME_MILLIS;
-    while (_periodicTaskInProgress && millisToWait > 0) {
-      try {
-        long thisWait = 1000;
-        if (millisToWait < thisWait) {
-          thisWait = millisToWait;
-        }
-        Thread.sleep(thisWait);
-        millisToWait -= thisWait;
-      } catch (InterruptedException e) {
-        LOGGER.info("Interrupted: Remaining wait time {} (out of {}) for task {}", millisToWait,
-            MAX_CONTROLLER_PERIODIC_TASK_STOP_TIME_MILLIS, getTaskName());
+    LOGGER.info("Processing {} tables in task: {}", numTables, _taskName);
+    C context = preprocess();
+    int numTablesProcessed = 0;
+    for (String tableNameWithType : tableNamesWithType) {
+      if (!isStarted()) {
+        LOGGER.info("Task: {} is stopped, early terminate the task", _taskName);
         break;
       }
-    }
-    LOGGER.info("Wait completed for task {}. Waited for {} ms. _periodicTaskInProgress = {}", getTaskName(),
-        MAX_CONTROLLER_PERIODIC_TASK_STOP_TIME_MILLIS - millisToWait, _periodicTaskInProgress);
-
-    stopTask();
-  }
-
-  /**
-   * Processes the task on the given tables.
-   *
-   * @param tableNamesWithType List of table names
-   */
-  protected void process(List<String> tableNamesWithType) {
-    if (!shouldStopPeriodicTask()) {
-
-      int numTablesProcessed = 0;
-      preprocess();
-
-      for (String tableNameWithType : tableNamesWithType) {
-        if (shouldStopPeriodicTask()) {
-          LOGGER.info("Skip processing table {} and all the remaining tables for task {}.", tableNameWithType,
-              getTaskName());
-          break;
-        }
-        try {
-          processTable(tableNameWithType);
-          numTablesProcessed++;
-        } catch (Exception e) {
-          exceptionHandler(tableNameWithType, e);
-        }
+      try {
+        processTable(tableNameWithType, context);
+      } catch (Exception e) {
+        LOGGER.error("Caught exception while processing table: {} in task: {}", tableNamesWithType, _taskName, e);
```

Comment:
why not keep exceptionHandler and call it here?

Reply:
Added back processTableExceptionHandler and pass in Context